src/storage/src/hummock/iterator/mod.rs

```diff
@@ -373,10 +379,93 @@ impl<'a, B: RustIteratorBuilder> FromRustIterator<'a, B> {
+                    user_key: UserKey {
+                        table_id: self.table_id,
+                        table_key: first_key,
+                    },
+                };
+                if first_full_key < key {
+                    // The semantic of `seek_fn` will ensure that `first_key` >= table_key of `key`.
+                    // At the beginning we have checked that `self.table_id` >= table_id of `key`.
+                    // Therefore, when `first_full_key` < `key`, the only possibility is that
+                    // `first_key` == table_key of `key`, and `self.table_id` == table_id of `key`,
+                    // the `self.epoch` < epoch of `key`.
+                    assert_eq!(first_key, key.user_key.table_key);
+                    match iter.next() {
+                        Some((next_key, next_value)) => {
+                            assert_gt!(next_key, first_key);
+                            self.iter =
+                                Some((RustIteratorOfBuilder::Seek(iter), next_key, next_value));
+                        }
+                        None => {
+                            self.iter = None;
+                        }
+                    }
+                } else {
+                    self.iter = Some((RustIteratorOfBuilder::Seek(iter), first_key, first_value));
+                }
+            }
+            None => {
+                self.iter = None;
+            }
+        }
+        Ok(())
+    }
+
+    async fn rev_seek_inner<'b>(&'b mut self, key: FullKey<&'b [u8]>) -> HummockResult<()> {
+        if self.table_id > key.user_key.table_id {
+            // returns None when the range of self.table_id must not include the given key
+            self.iter = None;
+            return Ok(());
+        }
+        if self.table_id < key.user_key.table_id {
+            return self.rewind().await;
+        }
+        let mut iter = B::seek(self.inner, key.user_key.table_key);
+        match iter.next() {
+            Some((first_key, first_value)) => {
+                let first_full_key = FullKey {
+                    epoch_with_gap: self.epoch,
+                    user_key: UserKey {
+                        table_id: self.table_id,
+                        table_key: first_key,
+                    },
+                };
+                if first_full_key > key {
+                    // The semantic of `seek_fn` will ensure that `first_key` <= table_key of `key`.
+                    // At the beginning we have checked that `self.table_id` <= table_id of `key`.
+                    // Therefore, when `first_full_key` > `key`, the only possibility is that
+                    // `first_key` == table_key of `key`, and `self.table_id` == table_id of `key`,
+                    // the `self.epoch` > epoch of `key`.
+                    assert_eq!(first_key, key.user_key.table_key);
+                }
+                self.iter = Some((RustIteratorOfBuilder::Seek(iter), first_key, first_value));
```

Comment:
Instead of checking `first_full_key > key`, I think we should check `first_key == key && key.epoch_with_gap > self.epoch` instead. When the condition is hit, we need to call `iter.next`. 

Example:
```
Memtable in epoch2:
 k1
 k2

rev iter seek(FullKey(k2, epoch3)) should point to k1 instead of k2 
because (k1, epoch2) < (k2, epoch3) < k2, epoch2)
```

Note that we cannot simply check `fisrt_full_key < key` because this won't guarantee that `first_key == key` under the semantic of reverse iter.

Reply:
We should add a test case for this as well.